### DISCUSSION TITLE
Trustless Notary Design - Ideation Sprint

### DISCUSSION THREAD

**Discussion by panges2** [filecoin-project/notary-governance#888] [Announcements]
## What is a Trustless Notary? 

With the recent introduction of the FEVM, developers are now empowered to compose smart contracts that seamlessly distribute tokens on-chain using logic capable of interacting with Filecoin chain state. In relation to Fil+, this innovation has unlocked a truly distinctive possibility. In recent months, the idea of a smart contract-based notary has gained traction among developers from different teams and contexts. We at the Fil+ team recognize this and want to further catalyze its development. As Fil+ scales, we believe that the transition towards a more fair and efficient method will be inevitable. The trustless notary is regarded as a popular way to facilitate this.

After reviewing a breadth of ideas, we have defined a "problem space" for the trustless notary that encapsulates all of the ideas under a unified general framework. This is done with the aim of establishing a standard understanding of what the Fil+ community defines as a "trustless notary". If the concept of a trustless notary remains unclear to you, or you're considering participating in the forthcoming design sprint, we strongly recommend reading the following article:

https://medium.com/filecoin-plus/ideation-trustless-notary-design-space-guidelines-bc21f6d9d5f2

Recognizing the potential significance of this project and the interest it has been receiving, we are excited to announce the trustless notary Ideation Sprint! We understand that our network resources are not infinite and that there's a variety of ways to implement this trustless notary, so the purpose of this ideation sprint is to aggregate and refine the community's top-tier ideas. This article holds all the information of how to partake in the sprint, as well as its requirements, judging criteria, and structure.

## Criteria 

- **Incentive Design:** The trustless notary should include an incentive structure that encourages participants to maintain the system's integrity. It should also reward users for "good behavior", fostering a fair and efficient ecosystem. For example, some possible ways to create an incentive structure are:
    - Having a market for clients paying notaries to do due-diligence on their applications
    - Notaries staking collateral to both incentivize and ensure accountability.
    - SPs paying verified clients for verified deals

- **Decision-Making Efficiency:** One of the major challenges with the existing LDN notary system is the extensive waiting periods for clients while notaries conduct due diligence and approve their applications. We will prioritize a trustless notary implementation that significantly reduces this decision-making time. Note that this doesn't discourage the inclusion of human notaries in the decision-making process. Instead, it paves the way for innovative solutions like accountability systems for human notaries.

- **Trustlessness:** The trustless notary should be decentralized, eliminating the need for intermediaries and minimizing trust-based risks. Decision-making processes should be transparently encoded in the smart contract and should rely on individual human notaries. An effective measure of a smart contract's trustlessness could be quantified by determining the minimal number of human actors required to deviate in order to exploit the system.

- **Technical Security:** The trustless notary should be resilient against cyber threats and any potential breaches. It should be developed in accordance with high-standard smart contract frameworks that prevent common security issues, such as reentrancy attacks. A good place to start is the [open zepplin](https://docs.openzeppelin.com/contracts/4.x/api/security) framework. If your submission gets community backing and is chosen, we may provide additional support through collaboration with external auditing teams.

- **Technical Efficiency:** The contract should be gas-efficient with minimal state size and limited state changes during interaction. Functions should not iterate over extensive states, and complexity should be reduced as much as possible. It should be designed to handle high transaction volumes without compromising speed or user experience. Further support may be provided through collaboration with expert smart contract development teams if we decide to go forward with the submission.

- **Permissionless Nature:** To maintain the ethos of decentralization, the trustless notary should be open to all individuals seeking to obtain datacap. It should not have unnecessary barriers to entry, promoting widespread adoption and composability with other smart contracts.

- **Fairness:** The smart contract logic should be fair in the issuing of datacap. Bias from individual human actors should be reduced to a minimum. Once implemented, an efficient and fair governance process should be in place to manage changes to the datacap issuance logic.
 
- **Feasibility:** The proposed notary should be technically and economically viable. It should be capable of being implemented within the existing infrastructure without too much costs and must be maintainable in the long term.

## Submission guidelines and timeline 

**Idea Design Submission Guidelines:**
- Ideas can be entered anytime during the submission window as comments in the discussion below. We encourage you to also link any relevant resources (docs, github repos etc.)
- Submitted ideas can be refined by community members via replying to those comments. 
- At the end of the submission period, ideas will be reviewed, and the top three design ideas will be shortlisted for further review by the Fil+ team and other core developers. These standout designs will then be eligible for potential support, which could come in the form of grant facilitation or the provision of engineering development assistance, either from our internal team or from external resources.

**Timeline:**
- Idea Submission Period: June 1 - June 30. This timeframe may be extended if more time is required.
- Idea Review Period: July 1 - 5.
- Discussion with top 3 submissions: July 6 - 15. During this period, in-depth discussions will take place to further refine the selected designs as well as prepare them for further advancement.
 

## Questions
If you have questions about this ideation sprint, please contact @Philippe Pangestu or @Kevin Z on the Filecoin Foundation Slack.

**Comment by The-Wayvy** [comment]
This is the most important thing that needs to happen in Filecoin.

We **can not** have a bureaucracy of humans handing out freebies to people whose data they like. 

All incentives **must** be written into the protocol.

We already have 88 notaries. The further we go down this path the more our network will be corrupted.

Verifiable, Permissionless, Immutable 

Nothing more nothing less.

UPDATE

Trustless notaries aren't happening.

**Comment by spaceT9** [comment]
The key mechanism is about staking for punishment or rewards?  Do I understand correctly?

**Comment by Fatman13** [comment]
> - Having a market for clients paying notaries to do due-diligence on their applications
> - Notaries staking collateral to both incentivize and ensure accountability.
> - SPs paying verified clients for verified deals

How would this ties into getting more paid deals for miners? It seems that it just encourages miners to reap the 10x QAP rewards.

> Governance can be structured to change the business logic of the smart contract, requiring multiple parties to agree before any changes are made to how datacap is distributed. This can help prevent collusion and ensure fairness.

Would there be a DAO token issued? I wouldn't personally trust any governance body if they have no skin in the game.

**Reply by Destore2023** [reply]
> > * Having a market for clients paying notaries to do due-diligence on their applications
> > * Notaries staking collateral to both incentivize and ensure accountability.
> > * SPs paying verified clients for verified deals
> 
> How would this ties into getting more paid deals for miners? It seems that it just encourages miners to reap the 10x QAP rewards.
> 
> > Governance can be structured to change the business logic of the smart contract, requiring multiple parties to agree before any changes are made to how datacap is distributed. This can help prevent collusion and ensure fairness.
> 
> Would there be a DAO token issued? I wouldn't personally trust any governance body if they have no skin in the game.

Hi @Fatman13 , you can also take a look here,  https://medium.com/filecoin-plus/ideation-trustless-notary-design-space-guidelines-bc21f6d9d5f2
